File: platforms/ios/Dezel/DezelTests/Core/JavaScriptClassTest.swift
import XCTest
import Dezel

class JavaScriptClassTest: XCTestCase {

	private var context: JavaScriptContext = JavaScriptContext()

	override func setUp() {
		super.setUp()
		self.context.dispose()
		self.context = JavaScriptContext()
	}

	func testJavaScriptClass() {

		var didCallConstructor = false
		var didCallFunction = false
		var didCallGetter = false
		var didCallSetter = false

		JavaScriptTestClass.testConstructorCallback = { callback in
			didCallConstructor = true
		}

		JavaScriptTestClass.testJavaScriptFunctionCallback = { callback in
			didCallFunction = true
		}

		JavaScriptTestClass.testJavaScriptGetterCallback = { callback in
			didCallGetter = true
		}

		JavaScriptTestClass.testJavaScriptSetterCallback = { callback in
			didCallSetter = true
		}

		let obj = self.constructJavaScriptClass(JavaScriptTestClass.self)

		obj.callMethod("testFunction")
		obj.property("testProperty", number: 10)
		obj.property("testProperty").dispose()
		obj.dispose()

		XCTAssertTrue(didCallConstructor)
		XCTAssertTrue(didCallFunction)
		XCTAssertTrue(didCallGetter)
		XCTAssertTrue(didCallSetter)
	}

	internal func constructJavaScriptClass(_ cls: AnyClass) -> JavaScriptValue {
		let obj = self.context.createReturnValue()
		let constructor = self.context.createClass(cls);
		constructor.construct(nil, result: obj)
		constructor.dispose()
		return obj;
	}
}

File: platforms/android/Dezel/dezel/src/androidTest/java/ca/logaritm/dezel/core/JavaScriptClassTest.kt
package ca.logaritm.dezel.core

import ca.logaritm.dezel.classes.ClassTemplateTestClass
import org.junit.Assert
import org.junit.Before
import org.junit.Test

class JavaScriptClassTest {

	private var context: JavaScriptContext = JavaScriptContext()

	@Before
	fun beforeTest() {
		this.context.dispose()
		this.context = JavaScriptContext()
	}

	@Test
	fun testClassTemplate() {

		var didCallConstructor = false
		var didCallFunction = false
		var didCallGetter = false
		var didCallSetter = false

		ClassTemplateTestClass.testConstructorCallback = fun(callback: JavaScriptFunctionCallback) {
			didCallConstructor = true
		}

		ClassTemplateTestClass.testFunctionCallback = fun(callback: JavaScriptFunctionCallback) {
			didCallFunction = true
		}

		ClassTemplateTestClass.testPropertyGetterCallback = fun(callback: JavaScriptGetterCallback) {
			didCallGetter = true
		}

		ClassTemplateTestClass.testPropertySetterCallback = fun(callback: JavaScriptSetterCallback) {
			didCallSetter = true
		}

		val obj = this.construct(ClassTemplateTestClass::class.java)
		obj.callMethod("testFunction")
		obj.property("testProperty", 10.0)
		obj.property("testProperty").dispose()
		obj.dispose()

		Assert.assertTrue(didCallConstructor)
		Assert.assertTrue(didCallFunction)
		Assert.assertTrue(didCallGetter)
		Assert.assertTrue(didCallSetter)
	}

	//--------------------------------------------------------------------------
	// Private API
	//--------------------------------------------------------------------------

	fun construct(type: Class<*>): JavaScriptValue {
		val target = context.createReturnValue()
		val constructor = context.createClass(type)
		constructor.construct(null, target)
		constructor.dispose()
		return target
	}
}
